Mathlib/RingTheory/TwoSidedIdeal.lean

```diff
@@ -0,0 +1,227 @@
+/-
+Copyright (c) 2024 Jujian Zhang. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Jujian Zhang
+-/
+
+import Mathlib.RingTheory.Congruence
+import Mathlib.RingTheory.Ideal.Basic
+import Mathlib.Algebra.Module.LinearMap.Basic
+import Mathlib.Algebra.Module.Opposites
+
+/-!
+# Two Sided Ideals
+
+In this file, for any `Ring R`, we reinterpret `I : RingCon R` as a two sided ideal of a ring.
+
+## Notes
+`SetLike (RingCon R) R` makes sense for any `NonUnitalNonAssocRing R`.
+But later for the `module` part, we will assume `R` is a ring.
+
+-/
+
+section ideal
+
+namespace RingCon
+
+section NonUnitalNonAssocRing
+
+variable {R : Type*} [NonUnitalNonAssocRing R] (I : RingCon R)
+
+instance : SetLike (RingCon R) R where
+  coe t := {r | t r 0}
+  coe_injective' t₁ t₂ (h : {x | _} = {x | _}) := by
+    refine RingCon.ext fun a b ↦ ⟨fun H ↦ ?_, fun H ↦ ?_⟩
+    · have H' : a - b ∈ {x | t₁ x 0} := sub_self b ▸  t₁.sub H (t₁.refl b)
+      rw [h] at H'
+      convert t₂.add H' (t₂.refl b) using 1 <;> abel
+    · have H' : a - b ∈ {x | t₂ x 0} := sub_self b ▸  t₂.sub H (t₂.refl b)
+      rw [← h] at H'
+      convert t₁.add H' (t₁.refl b) using 1 <;> abel
+
+lemma mem_iff (x : R) : x ∈ I ↔ I x 0 := Iff.rfl
+
+lemma rel_iff (x y : R) : I x y ↔ (x - y) ∈ I := by
+  rw [mem_iff]
+  constructor
+  · intro h; convert I.sub h (I.refl y); abel
+  · intro h; convert I.add h (I.refl y) <;> abel
+
+lemma le_iff (I J : RingCon R) : I ≤ J ↔ (I : Set R) ⊆ (J : Set R) := by
```

Comment:
This says the coercion to Set R is an [OrderEmbedding](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Order/Hom/Basic.html#OrderEmbedding); not sure how useful it is though

Reply:
added in [8629ec0](https://github.com/leanprover-community/mathlib4/pull/13401/commits/8629ec0d2f63d7c8ae524ec968b64f2de63ea10c)